src/main/java/org/folio/rest/impl/PermissionUtils.java

```diff
@@ -42,5 +62,49 @@ public static boolean equals(OkapiPermission okapiPerm, Permission perm) {
         && Objects.equals(okapiPerm.getDescription(), perm.getDescription())
         && Objects.equals(okapiPerm.getDisplayName(), perm.getDisplayName())
         && Objects.equals(okapiPerm.getSubPermissions(), perm.getSubPermissions());
-  }  
+  }
+
+  /**
+   * Purge deprecated permissions for a given tenant.
+   * 
+   * @param pgClient {@link PostgresClient}
+   * @param tenantId tenant id
+   * @return {@link Future} with {@link PermissionNameListObject}
+   */
+  public static Future<PermissionNameListObject> purgeDeprecatedPermissions(PostgresClient pgClient, String tenantId) {
+    Promise<PermissionNameListObject> promise = Promise.promise();
+    pgClient.startTx(tx -> {
+      try {
+        pgClient.select(tx, String.format(SELECT_DEPRECATED_PERMS, tenantId), res -> {
+          PermissionNameListObject permNames = new PermissionNameListObject();
+          permNames.setTotalRecords(0);
+          @SuppressWarnings("rawtypes")
+          List<Future> futures = new ArrayList<Future>();
+          res.result()
+            .forEach(row -> {
+              String id = row.getString("id");
+              String name = row.getString("name");
+              futures.add(Future.<RowSet<Row>>future(p -> pgClient.execute(tx, String.format(PURGE_DEPRECATED_SUB_PERMS, tenantId, id, name, id, name), p)));
+              futures.add(Future.<RowSet<Row>>future(p -> pgClient.execute(tx, String.format(PURGE_DEPRECATED_PERMS_USERS, tenantId, id, name, id, name), p)));
+              futures.add(Future.<RowSet<Row>>future(p -> pgClient.execute(tx, String.format(PURGE_DEPRECATED_PERMS, tenantId), p)));
+              permNames.getPermissionNames().add(name);
+              permNames.setTotalRecords(permNames.getTotalRecords() + 1);
+            });
+          CompositeFuture.all(futures)
+            .onSuccess(ignore -> {
+              pgClient.endTx(tx, done -> done.mapEmpty());
+              promise.complete(permNames);
```

Comment:
the done.map is redundant. It is desired that completion is not signalled before endTx/rollback is complete.
```suggestion
              pgClient.endTx(tx, done -> promise.complete(permNames));
```

Reply:
Make sense. Changed.